File: frontend/src/lib/invite-jwt.ts
/**
 * Invite-flow JWT (HS256).
 *
 * Phase 1 of the Chess Empire → Chesster onboarding arc. After a parent
 * passes the DOB gate (verify route), the server signs a short-lived JWT
 * containing the resolved student/branch context. The token is handed to
 * the Clerk sign-up page via `?invite=…`, and replayed back through the
 * Clerk webhook in Phase 2/3 to write the `external_student_id` linkage.
 *
 * Plumbed without a new dependency — `node:crypto` HMAC is sufficient for
 * HS256 and the JWT format is small. Mirrors the Python equivalent at
 * ``backend/services/invite_jwt.py`` which uses ``pyjwt``.
 *
 * Default TTL: 900 seconds (15 min). The webhook also enforces single-use
 * via a `consumed_at` write in Phase 2/3.
 */
import 'server-only';
import { createHash, createHmac, timingSafeEqual } from 'node:crypto';

export const INVITE_JWT_TTL_SECONDS = 15 * 60;

export type MemberType = 'student' | 'coach';

export interface InviteJwtPayload {
  /** Chess Empire student (or coach) UUID. */
  student_id: string;
  /** Chess Empire branch UUID (mirrors the token's branch). */
  branch_id: string;
  /** Chesster `branch_invite_tokens.id` — used by the webhook for audit. */
  branch_token_id: string;
  /** Chesster `organizations.id`. */
  org_id: string;
  /**
   * Member role for the linked `organization_members` row. Optional and
   * absent for legacy student tokens — verify normalizes a missing claim to
   * `'student'` so pre-existing tokens stay valid (back-compat).
   */
  member_type?: MemberType;
}

interface InviteJwtClaims extends InviteJwtPayload {
  /** Always populated after verify — normalized to `'student'` when absent. */
  member_type: MemberType;
  iat: number;
  exp: number;
}

export class InviteJwtError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'InviteJwtError';
  }
}

function getSecret(): string {
  const secret = process.env.INVITE_JWT_SECRET;
  if (!secret) {
    throw new InviteJwtError('INVITE_JWT_SECRET not configured');
  }
  return secret;
}

function b64urlEncode(buf: Buffer): string {
  return buf.toString('base64').replace(/=+$/, '').replace(/\+/g, '-').replace(/\//g, '_');
}

function b64urlDecode(str: string): Buffer {
  const pad = str.length % 4 === 0 ? '' : '='.repeat(4 - (str.length % 4));
  return Buffer.from(str.replace(/-/g, '+').replace(/_/g, '/') + pad, 'base64');
}

function sign(headerAndPayload: string, secret: string): string {
  return b64urlEncode(createHmac('sha256', secret).update(headerAndPayload).digest());
}

export function signInviteJwt(
  payload: InviteJwtPayload,
  ttlSeconds: number = INVITE_JWT_TTL_SECONDS,
  nowSeconds: number = Math.floor(Date.now() / 1000),
): string {
  const secret = getSecret();
  const header = { alg: 'HS256', typ: 'JWT' };
  // member_type stays optional here — an absent claim signs a legacy-shaped
  // student token; verify normalizes a missing claim back to 'student'.
  const claims: InviteJwtPayload & { iat: number; exp: number } = {
    ...payload,
    iat: nowSeconds,
    exp: nowSeconds + ttlSeconds,
  };
  const encodedHeader = b64urlEncode(Buffer.from(JSON.stringify(header)));
  const encodedPayload = b64urlEncode(Buffer.from(JSON.stringify(claims)));
  const signingInput = `${encodedHeader}.${encodedPayload}`;
  return `${signingInput}.${sign(signingInput, secret)}`;
}

export function verifyInviteJwt(
  token: string,
  nowSeconds: number = Math.floor(Date.now() / 1000),
): InviteJwtClaims {
  const secret = getSecret();
  const parts = token.split('.');
  if (parts.length !== 3) {
    throw new InviteJwtError('Malformed token');
  }
  const [encodedHeader, encodedPayload, encodedSig] = parts;
  const expected = sign(`${encodedHeader}.${encodedPayload}`, secret);
  const expectedBuf = Buffer.from(expected);
  const sigBuf = Buffer.from(encodedSig);
  if (expectedBuf.length !== sigBuf.length || !timingSafeEqual(expectedBuf, sigBuf)) {
    throw new InviteJwtError('Bad signature');
  }
  let claims: InviteJwtClaims;
  try {
    claims = JSON.parse(b64urlDecode(encodedPayload).toString('utf8')) as InviteJwtClaims;
  } catch {
    throw new InviteJwtError('Malformed payload');
  }
  if (typeof claims.exp !== 'number' || claims.exp < nowSeconds) {
    throw new InviteJwtError('Token expired');
  }
  if (!claims.student_id || !claims.branch_id || !claims.branch_token_id || !claims.org_id) {
    throw new InviteJwtError('Missing required claim');
  }
  // Back-compat: legacy tokens carry no member_type — treat them as students.
  claims.member_type = claims.member_type === 'coach' ? 'coach' : 'student';
  return claims;
}

/**
 * Deterministic sha256 hex of the raw JWT string.
 *
 * Mirror of Python `jwt_jti_hash` in `backend/services/invite_jwt.py`. Used by
 * the `user.created` webhook to record single-use consumption in
 * `invite_jwts_consumed` without ever writing the raw token to the DB.
 * Same token → same hash → PK conflict → replay is a no-op.
 */
export function jwtJtiHash(token: string): string {
  return createHash('sha256').update(token, 'utf8').digest('hex');
}

File: backend/services/invite_jwt.py
"""
Invite-flow JWT (HS256, Python mirror).

Phase 1 of the Chess Empire → Chesster onboarding arc. Mirror of
``frontend/src/lib/invite-jwt.ts``. Used by the Clerk webhook completion
path (Phase 2/3) to verify the JWT a parent picked up on the sign-up page,
before writing the ``external_student_id`` linkage onto the freshly created
``organization_members`` row.

Uses ``pyjwt`` (already in requirements.txt for Clerk JWT verification).
15-minute TTL by default. Single-use enforcement (``consumed_at``) is the
webhook's responsibility, not this module's.
"""

from __future__ import annotations

import hashlib
import os
import time
from typing import TypedDict

import jwt
from jwt.exceptions import InvalidTokenError, ExpiredSignatureError

INVITE_JWT_TTL_SECONDS = 15 * 60
INVITE_JWT_ALGORITHM = 'HS256'

REQUIRED_CLAIMS = ('student_id', 'branch_id', 'branch_token_id', 'org_id')


class InviteJwtPayload(TypedDict):
    student_id: str
    branch_id: str
    branch_token_id: str
    org_id: str


class InviteJwtClaims(InviteJwtPayload):
    iat: int
    exp: int


class InviteJwtError(Exception):
    """Raised on any signing/verification failure."""


def _get_secret() -> str:
    secret = os.getenv('INVITE_JWT_SECRET', '')
    if not secret:
        raise InviteJwtError('INVITE_JWT_SECRET not configured')
    return secret


def sign_invite_jwt(
    payload: InviteJwtPayload,
    ttl_seconds: int = INVITE_JWT_TTL_SECONDS,
    now: int | None = None,
) -> str:
    """Sign a 15-min HS256 JWT carrying the invite context."""
    now_seconds = int(time.time()) if now is None else now
    claims = {
        **payload,
        'iat': now_seconds,
        'exp': now_seconds + ttl_seconds,
    }
    return jwt.encode(claims, _get_secret(), algorithm=INVITE_JWT_ALGORITHM)


def verify_invite_jwt(token: str, now: int | None = None) -> InviteJwtClaims:
    """Verify signature + expiry. Raises InviteJwtError on any failure."""
    secret = _get_secret()
    # When the caller supplies `now` (tests), do exp validation manually so we
    # aren't locked to wall-clock time. Otherwise let pyjwt enforce it.
    verify_exp = now is None
    try:
        decoded = jwt.decode(
            token,
            secret,
            algorithms=[INVITE_JWT_ALGORITHM],
            options={'require': ['exp'], 'verify_exp': verify_exp},
            leeway=0,
        )
    except ExpiredSignatureError as exc:
        raise InviteJwtError('Token expired') from exc
    except InvalidTokenError as exc:
        raise InviteJwtError(str(exc) or 'Invalid token') from exc

    if now is not None:
        exp = decoded.get('exp')
        if not isinstance(exp, int) or exp < now:
            raise InviteJwtError('Token expired')

    for claim in REQUIRED_CLAIMS:
        if not decoded.get(claim):
            raise InviteJwtError(f'Missing required claim: {claim}')

    return decoded  # type: ignore[return-value]


def jwt_jti_hash(token: str) -> str:
    """Deterministic sha256 hex of the raw JWT string.

    Used by the ``user.created`` webhook to record single-use consumption in
    ``invite_jwts_consumed`` without ever writing the raw token to the DB.
    Same token → same hash → PK conflict → replay is a no-op.
    """
    return hashlib.sha256(token.encode('utf-8')).hexdigest()
